File: tests/python/test_monitor.py
from zeus_apple_silicon import AppleEnergyMonitor, AppleEnergyMetrics
import pytest


def dummy_work(limit):
    """Dummy work to exercise the machine a little bit."""
    x = 0
    for i in range(limit):
        for j in range(10000):
            x += j


def test_energy_monitor():
    """Simulate a well-formulated usage of the energy monitor."""
    monitor = AppleEnergyMonitor()
    assert isinstance(monitor, AppleEnergyMonitor)

    monitor.begin_window("test1")
    dummy_work(1000)
    result = monitor.end_window("test1")

    assert isinstance(result, AppleEnergyMetrics)

    # Regardless of the specific Apple chip, cpu_total_mj and gpu_mj should generally be non-None.
    assert result.cpu_total_mj is not None
    assert result.gpu_mj is not None


# TODO: add more tests

File: tests/cpp/test_monitor.cpp
#include "energy.hpp"
#include "mock_ioreport.hpp"
#include <cassert>
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>

/* Implemented by mock infrastructure. */
void set_mock_sample(const std::unordered_map<std::string, int64_t>& data);
void reset_mock_state();

void test_one_interval()
{
    std::cout << "* Running test_one_interval\n";
    Mocker mocker;

    std::unordered_map<std::string, int64_t> data1 = {
        { "CPU Energy", 10000000 }, // in mJ
        { "GPU Energy", 1000000 }, // in nJ
    };
    mocker.push_back_sample(data1);

    std::unordered_map<std::string, int64_t> data2 = {
        { "CPU Energy", 55000000 },
        { "GPU Energy", 3000000 },
    };
    mocker.push_back_sample(data2);

    AppleEnergyMonitor monitor;

    monitor.begin_window("test");
    AppleEnergyMetrics result = monitor.end_window("test");

    assert(result.cpu_total_mj.value() == 45000000);
    assert(result.gpu_mj.value() == 2);
}

int main()
{
    std::cout << "--- Starting tests ---\n";

    test_one_interval();

    std::cout << "All tests passed.\n";
}
